File: scribbleshare-room/src/main/java/net/stzups/scribbleshare/room/server/websocket/protocol/server/messages/ServerMessageCanvasInsert.java
package net.stzups.scribbleshare.room.server.websocket.protocol.server.messages;

import io.netty.buffer.ByteBuf;
import net.stzups.scribbleshare.data.objects.canvas.CanvasInsert;
import net.stzups.scribbleshare.data.objects.canvas.object.CanvasObjectType;
import net.stzups.scribbleshare.room.server.websocket.protocol.server.ServerMessage;
import net.stzups.scribbleshare.room.server.websocket.protocol.server.ServerMessageType;

import java.util.Map;

public class ServerMessageCanvasInsert extends ServerMessage {
    Map<CanvasObjectType, CanvasInsert[]> canvasInsertsMap;

    public ServerMessageCanvasInsert(Map<CanvasObjectType, CanvasInsert[]> canvasInsertsMap) {
        super(ServerMessageType.CANVAS_INSERT);
        this.canvasInsertsMap = canvasInsertsMap;
    }

    public void serialize(ByteBuf byteBuf) {
        byteBuf.writeByte((byte) canvasInsertsMap.size());
        for (Map.Entry<CanvasObjectType, CanvasInsert[]> entry : canvasInsertsMap.entrySet()) {
            byteBuf.writeByte((byte) entry.getKey().getId());
            byteBuf.writeByte((byte) entry.getValue().length);
            for (CanvasInsert canvasInsert : entry.getValue()) {
                canvasInsert.serialize(byteBuf);
            }
        }
    }
}

File: scribbleshare-frontend/src/scripts/protocol/server/messages/ServerMessageCanvasInsert.js
import ServerMessageType from "../ServerMessageType.js";
import ServerMessage from "../ServerMessage.js";
import CanvasInsert from "../../../canvas/CanvasInsert.js";

export default class ServerMessageCanvasInsert extends ServerMessage {
    constructor(reader) {
        super(ServerMessageType.CANVAS_INSERT);
        this.canvasInsertsMap = new Map();
        let length = reader.readUint8();
        for (let i = 0; i < length; i++) {
            let canvasObjectType = reader.readUint8();
            let canvasInserts = [];
            let lengthJ = reader.readUint8();
            for (let j = 0; j < lengthJ; j++) {
                canvasInserts.push(new CanvasInsert(canvasObjectType, reader));
            }
            this.canvasInsertsMap.set(canvasObjectType, canvasInserts);
        }
    }
}